seed/seedwriter/seed20.go

```diff
@@ -0,0 +1,316 @@
+
+func (tr *tree20) writeAssertions(db asserts.RODatabase, modelRefs []*asserts.Ref, snapsFromModel []*SeedSnap, extraSnaps []*SeedSnap) error {
+	assertsDir := filepath.Join(tr.systemDir, "assertions")
+	if err := os.MkdirAll(assertsDir, 0755); err != nil {
+		return err
+	}
+
+	writeRefs := func(fname string, refsGen func(stop <-chan struct{}) <-chan *asserts.Ref) error {
+		f, err := os.OpenFile(filepath.Join(assertsDir, fname), os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0644)
+		if err != nil {
+			return err
+		}
+		defer f.Close()
+
+		stop := make(chan struct{})
+		defer close(stop)
+		refs := refsGen(stop)
+
+		enc := asserts.NewEncoder(f)
+		for {
+			aRef := <-refs
+			if aRef == nil {
+				break
+			}
+			a, err := aRef.Resolve(db.Find)
+			if err != nil {
+				return fmt.Errorf("internal error: lost saved assertion")
+			}
+			if err := enc.Encode(a); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	pushRef := func(refs chan<- *asserts.Ref, ref *asserts.Ref, stop <-chan struct{}) bool {
+		select {
+		case refs <- ref:
+			return true
+		case <-stop:
+			// get unstuck if we error early
+			return false
+		}
+	}
+
+	modelRefsGen := func(model bool) func(stop <-chan struct{}) <-chan *asserts.Ref {
+		return func(stop <-chan struct{}) <-chan *asserts.Ref {
+			refs := make(chan *asserts.Ref)
+			go func() {
+				for _, aRef := range modelRefs {
+					isModel := aRef.Type == asserts.ModelType
+					if model && !isModel {
+						continue
+					}
+					if !model && isModel {
+						continue
+					}
+					if !pushRef(refs, aRef, stop) {
+						return
+					}
+				}
+				close(refs)
+			}()
+			return refs
+		}
+	}
+
+	if err := writeRefs("../model", modelRefsGen(true)); err != nil {
+		return err
+	}
+
+	if err := writeRefs("model-etc", modelRefsGen(false)); err != nil {
+		return err
+	}
+
+	snapsRefGen := func(snaps []*SeedSnap) func(stop <-chan struct{}) <-chan *asserts.Ref {
+		return func(stop <-chan struct{}) <-chan *asserts.Ref {
+			refs := make(chan *asserts.Ref)
+			go func() {
+				for _, sn := range snaps {
+					for _, aRef := range sn.ARefs {
+						if !pushRef(refs, aRef, stop) {
+							return
+						}
+					}
+				}
+				close(refs)
+			}()
+			return refs
+		}
+	}
+
+	if err := writeRefs("snaps", snapsRefGen(snapsFromModel)); err != nil {
+		return err
+	}
+
+	if len(extraSnaps) != 0 {
+		if err := writeRefs("extra-snaps", snapsRefGen(extraSnaps)); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+type auxInfo20 struct {
+	Private bool   `json:"private,omitempty"`
+	Contact string `json:"contact,omitempty"`
+}
+
+func (tr *tree20) writeMeta(snapsFromModel []*SeedSnap, extraSnaps []*SeedSnap) error {
+	// TODO|XXX: produce an options.yaml if needed (grade dangerous)
+
+	auxInfos := make(map[string]*auxInfo20)
+
+	addAuxInfos := func(seedSnaps []*SeedSnap) {
+		for _, sn := range seedSnaps {
+			if sn.Info.ID() != "" {
+				if sn.Info.Contact != "" || sn.Info.Private {
+					auxInfos[sn.Info.ID()] = &auxInfo20{
```

Comment:
we don't plan to support parallel installs from the seeds at this point.  Even if we did because the point of this information is from the store and it for store snaps, I would still use some snap-id based key, plus something more.